feat: be generic, support any SQLx database

Fishrock123 opened the pull request:
This changes a lot of this crate to be generic over `sqlx::Database`,
and also changes the naming of everything to be "SQLx*".

Additionally, this now has a greatly simplified `ConnectionWrapInner`
implementation via `Deref` and `DerefMut`! Thanks to @mehcode for that!

----

Future improvements, either in this PR or in follow-ups include: 
- `tide_sqlx::PostgresMiddleware` (and etc) as aliases to e.g. `tide_sqlx::SQLxMiddleware<Postgres>` when the postgres feature is enabled.
- `tide_sqlx::PostgresRequestExt` (and etc) as shortcusts to allow `req.postgres_conn()` rather than `req.sqlx_conn::<Postgres>()`.

----

note to self: still need to remove the `"postgres"` and `"any"` features by default, only include in test runs

Fishrock123 committed:
feat: be generic, support any SQLx database

This changes a lot of this crate to be generic over `sqlx::Database`,
and also changes the naming of everything to be "SQLx*".

Additionally, this now has a greatly simplified `ConnectionWrapInner`
implementation via `Deref` and `DerefMut`! Thanks to @mehcode for that!

chrisdickinson commented on Cargo.toml:
```diff
@@ -42,7 +42,7 @@ futures-core = "0.3.8"
 [dependencies.sqlx]
 version = "0.4.1"
-features = ["postgres"]
+features = ["any"]
```
Is `"any"` a distinct feature flag for `sqlx`, or someting particular to rustc's feature support?

mehcode replied:
@chrisdickinson "any" is a feature flag for `sqlx` that turns on the "Any" database driver. That more closely maps to what JDBC does and allows for configuring the db driver at runtime instead of compile-time. 

---

I would recommend specifying no db/type feature flags in Cargo.toml and recommending people to use tide-sqlx like this:

```toml
sqlx = { version = "0.4", features = [ "postgres" ] }
tide-sqlx = "0.1"
```

As features are additive, it'll apply inside tide-sqlx as well.

chrisdickinson commented on src/lib.rs:
```diff
@@ -210,115 +215,67 @@ pub trait PostgresRequestExt<'req> {
+// postgres = [ "sqlx/postgres"]
+// mysql = [ "sqlx/mysql" ]
+// sqlite = [ "sqlx/sqlite" ]
+// mssql = [ "sqlx/mssql" ]
```
Ah, errant comment?

chrisdickinson reviewed (approved):
A question and a comment about a comment block aside, this looks great and is a wonderful simplification. LGTM!

Fishrock123 committed:
feat: nice compile error for tests without postgres

Fishrock123 committed:
fix: correct cargo features

Fishrock123 committed:
docs: update documentation for generic databases

Fishrock123 committed:
refactor: reorganize ConnectionWrapInner impls

Fishrock123 committed:
fix: avoid unsafe impl Send + Sync

Tide extensions must be `Send + Sync + 'static`.

Unfortunately this means we cannot support SQLx's sqlite connection, as it is `!Sync` for some reason:
https://docs.rs/sqlx/0.4.1/sqlx/struct.SqliteConnection.html#impl-Sync
